File: src/lib/aethos/astrology/aspects.ts
import { normalizeLongitude } from "./zodiac-format";
import type { Aspect, AspectType, CelestialPosition, OrbConfig } from "./types";

export const ASPECT_ANGLES: Record<AspectType, number> = {
  conjunction: 0,
  sextile: 60,
  square: 90,
  trine: 120,
  opposition: 180
};

export const DEFAULT_ORBS: Required<OrbConfig> = {
  conjunction: 8,
  sextile: 4,
  square: 6,
  trine: 6,
  opposition: 8
};

export function angularDistance(a: number, b: number): number {
  const diff = Math.abs(normalizeLongitude(a) - normalizeLongitude(b));
  return diff > 180 ? 360 - diff : diff;
}

export function calculateAspect(
  bodyA: CelestialPosition,
  bodyB: CelestialPosition,
  orbConfig: OrbConfig = DEFAULT_ORBS
): Aspect | null {
  const distance = angularDistance(bodyA.longitude, bodyB.longitude);
  const matches = Object.entries(ASPECT_ANGLES)
    .map(([type, angle]) => ({
      type: type as AspectType,
      angle,
      orb: Math.abs(distance - angle),
      maxOrb: orbConfig[type as AspectType] ?? DEFAULT_ORBS[type as AspectType]
    }))
    .filter((candidate) => candidate.orb <= candidate.maxOrb)
    .sort((a, b) => a.orb - b.orb);

  const match = matches[0];
  if (!match) return null;

  return {
    id: `${bodyA.body}-${bodyB.body}-${match.type}`,
    bodyA: bodyA.body,
    bodyB: bodyB.body,
    type: match.type,
    orb: Number(match.orb.toFixed(3)),
    exactAngle: match.angle,
    applying: Math.abs(bodyA.speed.longitudePerDay) >= Math.abs(bodyB.speed.longitudePerDay)
  };
}

export function calculateAspects(positions: CelestialPosition[], orbConfig?: OrbConfig): Aspect[] {
  const aspects: Aspect[] = [];
  for (let i = 0; i < positions.length; i += 1) {
    for (let j = i + 1; j < positions.length; j += 1) {
      const aspect = calculateAspect(positions[i], positions[j], orbConfig);
      if (aspect) aspects.push(aspect);
    }
  }
  return aspects;
}

File: services/calculation/app/calculations/aspects.py
from app.calculations.zodiac import normalize_longitude
from app.models.astrology import Aspect, AspectType, CelestialPosition
from app.models.requests import OrbConfig

ASPECT_ANGLES: dict[AspectType, float] = {
    AspectType.conjunction: 0,
    AspectType.semisextile: 30,
    AspectType.sextile: 60,
    AspectType.square: 90,
    AspectType.trine: 120,
    AspectType.quincunx: 150,
    AspectType.opposition: 180,
}


def shortest_angular_distance(a: float, b: float) -> float:
    diff = abs(normalize_longitude(a) - normalize_longitude(b))
    return 360 - diff if diff > 180 else diff


def calculate_aspect(body_a: CelestialPosition, body_b: CelestialPosition, orb_config: OrbConfig) -> Aspect | None:
    separation = shortest_angular_distance(body_a.longitude, body_b.longitude)
    candidates: list[tuple[AspectType, float, float, float]] = []
    for aspect_type, exact_angle in ASPECT_ANGLES.items():
        orb = abs(separation - exact_angle)
        max_orb = orb_config.for_aspect(aspect_type)
        if orb <= max_orb:
            candidates.append((aspect_type, exact_angle, orb, max_orb))
    if not candidates:
        return None
    aspect_type, exact_angle, orb, max_orb = sorted(candidates, key=lambda item: item[2])[0]
    exactness = max(0.0, 1.0 - orb / max_orb) if max_orb else 1.0
    return Aspect(
        bodyA=body_a.body.value,
        bodyB=body_b.body.value,
        aspectType=aspect_type,
        exactAngle=exact_angle,
        actualSeparation=round(separation, 6),
        orb=round(orb, 6),
        maximumOrb=max_orb,
        withinOrb=True,
        applyingSeparating="applying" if abs(body_a.speedLongitude) >= abs(body_b.speedLongitude) else "separating",
        exactnessScore=round(exactness, 6),
    )


def calculate_aspects(positions: list[CelestialPosition], orb_config: OrbConfig) -> list[Aspect]:
    aspects: list[Aspect] = []
    for index, body_a in enumerate(positions):
        for body_b in positions[index + 1 :]:
            aspect = calculate_aspect(body_a, body_b, orb_config)
            if aspect:
                aspects.append(aspect)
    return aspects
